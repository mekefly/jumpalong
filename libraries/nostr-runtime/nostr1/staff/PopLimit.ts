import EventEmitter from "events";
import { Event } from "nostr-tools";
import { createStaffFactory } from "./Staff1";

export default createStaffFactory()((limit: number) => {
  const eventEmiter = new EventEmitter();
  return {
    push(event: Event, eventList) {
      if (eventList.length >= limit) {
        const popEvent = eventList.shift();
        eventEmiter.emit("limit-pop", popEvent);
      }
    },
    feat: {
      onLimitPop(callBack: (e: Event) => void) {
        eventEmiter.on("limit-pop", callBack);
      },
    },
  };
});
