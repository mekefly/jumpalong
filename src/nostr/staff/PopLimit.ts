import EventEmitter from "events";
import { Event } from "nostr-tools";
import { createStaffFactory } from "./Staff";

export default createStaffFactory()((limit: number) => {
  const eventEmiter = new EventEmitter();
  return {
    afterPush(event: Event, eventList) {
      if (eventList.length > limit) {
        const popEvent = eventList.pop();
        eventEmiter.emit("limit-pop", popEvent);
      }
    },
    feat: {
      onLimitPop(callBack: (e: Event) => void) {
        eventEmiter.on("limit-pop", callBack);
      },
    },
  };
});
