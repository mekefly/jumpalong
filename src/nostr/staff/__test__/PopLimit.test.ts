import { Event } from "nostr-tools";
import PopLimit from "../PopLimit";

it("PopLimit", () => {
  const state = PopLimit(10);
  expect(state).toMatchInlineSnapshot(`
    {
      "feat": {
        "onLimitPop": [Function],
      },
      "push": [Function],
    }
  `);
});
it("PopLimit:Push", () => {
  const state = PopLimit(10);
  const list: Event[] = [];

  const mockLine = {
    beltline: {
      feat: {},
      getList() {
        return list;
      },
    },
  } as any;
  for (const v of Array.from({ length: 30 }, (itme, index) => index)) {
    const event = {
      id: v.toString(),
      content: v.toString(),
      kind: 3,
      pubkey: "p",
      sig: "33",
      tags: [],
      created_at: 1683187195 + v,
    };
    state.push.call(mockLine, event, list);
    list.push(event);
  }
  expect(list).toMatchInlineSnapshot(`
    [
      {
        "content": "20",
        "created_at": 1683187215,
        "id": "20",
        "kind": 3,
        "pubkey": "p",
        "sig": "33",
        "tags": [],
      },
      {
        "content": "21",
        "created_at": 1683187216,
        "id": "21",
        "kind": 3,
        "pubkey": "p",
        "sig": "33",
        "tags": [],
      },
      {
        "content": "22",
        "created_at": 1683187217,
        "id": "22",
        "kind": 3,
        "pubkey": "p",
        "sig": "33",
        "tags": [],
      },
      {
        "content": "23",
        "created_at": 1683187218,
        "id": "23",
        "kind": 3,
        "pubkey": "p",
        "sig": "33",
        "tags": [],
      },
      {
        "content": "24",
        "created_at": 1683187219,
        "id": "24",
        "kind": 3,
        "pubkey": "p",
        "sig": "33",
        "tags": [],
      },
      {
        "content": "25",
        "created_at": 1683187220,
        "id": "25",
        "kind": 3,
        "pubkey": "p",
        "sig": "33",
        "tags": [],
      },
      {
        "content": "26",
        "created_at": 1683187221,
        "id": "26",
        "kind": 3,
        "pubkey": "p",
        "sig": "33",
        "tags": [],
      },
      {
        "content": "27",
        "created_at": 1683187222,
        "id": "27",
        "kind": 3,
        "pubkey": "p",
        "sig": "33",
        "tags": [],
      },
      {
        "content": "28",
        "created_at": 1683187223,
        "id": "28",
        "kind": 3,
        "pubkey": "p",
        "sig": "33",
        "tags": [],
      },
      {
        "content": "29",
        "created_at": 1683187224,
        "id": "29",
        "kind": 3,
        "pubkey": "p",
        "sig": "33",
        "tags": [],
      },
    ]
  `);
});

it("PopLimit:onLimitPop", () => {
  const state = PopLimit(10);
  const list: Event[] = [];

  const mockLine = {
    beltline: {
      feat: {},
      getList() {
        return list;
      },
    },
  } as any;
  state.feat.onLimitPop.call(mockLine, (event) => {
    expect(event).toMatchSnapshot();
  });
  for (const v of Array.from({ length: 15 }, (itme, index) => index)) {
    const event = {
      id: v.toString(),
      content: v.toString(),
      kind: 3,
      pubkey: "p",
      sig: "33",
      tags: [],
      created_at: 1683187195 + v,
    };
    state.push.call(mockLine, event, list);
    list.push(event);
  }
});
