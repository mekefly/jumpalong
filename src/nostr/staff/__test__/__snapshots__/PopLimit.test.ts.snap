// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PopLimit:Push 2`] = `
{
  "content": "0",
  "created_at": 1683187195,
  "id": "0",
  "kind": 3,
  "pubkey": "p",
  "sig": "33",
  "tags": [],
}
`;

exports[`PopLimit:Push 3`] = `
{
  "content": "1",
  "created_at": 1683187196,
  "id": "1",
  "kind": 3,
  "pubkey": "p",
  "sig": "33",
  "tags": [],
}
`;

exports[`PopLimit:Push 4`] = `
{
  "content": "2",
  "created_at": 1683187197,
  "id": "2",
  "kind": 3,
  "pubkey": "p",
  "sig": "33",
  "tags": [],
}
`;

exports[`PopLimit:Push 5`] = `
{
  "content": "3",
  "created_at": 1683187198,
  "id": "3",
  "kind": 3,
  "pubkey": "p",
  "sig": "33",
  "tags": [],
}
`;

exports[`PopLimit:Push 6`] = `
{
  "content": "4",
  "created_at": 1683187199,
  "id": "4",
  "kind": 3,
  "pubkey": "p",
  "sig": "33",
  "tags": [],
}
`;

exports[`PopLimit:onLimitPop 1`] = `
{
  "content": "0",
  "created_at": 1683187195,
  "id": "0",
  "kind": 3,
  "pubkey": "p",
  "sig": "33",
  "tags": [],
}
`;

exports[`PopLimit:onLimitPop 2`] = `
{
  "content": "1",
  "created_at": 1683187196,
  "id": "1",
  "kind": 3,
  "pubkey": "p",
  "sig": "33",
  "tags": [],
}
`;

exports[`PopLimit:onLimitPop 3`] = `
{
  "content": "2",
  "created_at": 1683187197,
  "id": "2",
  "kind": 3,
  "pubkey": "p",
  "sig": "33",
  "tags": [],
}
`;

exports[`PopLimit:onLimitPop 4`] = `
{
  "content": "3",
  "created_at": 1683187198,
  "id": "3",
  "kind": 3,
  "pubkey": "p",
  "sig": "33",
  "tags": [],
}
`;

exports[`PopLimit:onLimitPop 5`] = `
{
  "content": "4",
  "created_at": 1683187199,
  "id": "4",
  "kind": 3,
  "pubkey": "p",
  "sig": "33",
  "tags": [],
}
`;
